lib/srv/db/mysql/autousers.go

```diff
@@ -469,21 +516,55 @@ var (
 	//go:embed mysql_revoke_roles.sql
 	revokeRolesProcedure string
 
-	allProcedures = []struct {
-		name          string
-		createCommand string
-	}{
-		{
-			name:          revokeRolesProcedureName,
-			createCommand: revokeRolesProcedure,
-		},
-		{
-			name:          activateUserProcedureName,
-			createCommand: activateUserProcedure,
-		},
-		{
-			name:          deactivateUserProcedureName,
-			createCommand: deactivateUserProcedure,
-		},
+	//go:embed mariadb_activate_user.sql
+	mariadbActivateUserProcedure string
+	//go:embed mariadb_deactivate_user.sql
+	mariadbDeactivateUserProcedure string
+	//go:embed mariadb_revoke_roles.sql
+	mariadbRevokeRolesProcedure string
+
+	// allProcedureNames contains a list of all procedures required to setup
+	// auto-user provisioning. Note that order matters here as later procedures
+	// may depend on the earlier ones.
+	allProcedureNames = []string{
+		revokeRolesProcedureName,
+		activateUserProcedureName,
+		deactivateUserProcedureName,
+	}
+
+	// mysqlProcedures maps procedure names to the procedures used for MySQL.
+	mysqlProcedures = map[string]string{
+		activateUserProcedureName:   activateUserProcedure,
+		deactivateUserProcedureName: deactivateUserProcedure,
+		revokeRolesProcedureName:    revokeRolesProcedure,
+	}
+
+	// mariadbProcedures maps procedure names to the procedures used for MariaDB.
+	//
+	// MariaDB requires separate procedures from MySQL because:
+	// - Max length of username/role is 80/128 instead of 32.
+	// - MariaDB uses mysql.roles_mapping instead of mysql.role_edges.
+	//   Note that roles_mapping tracks both role assignments and role "owners".
+	//   "Owners" are tracked in rows with Admin_option = 'Y'.
+	// - MariaDB does not have built-in user attributes field. Instead, the
+	//   procedure will create an `user_attributes` table for tracking this. A
+	//   view `V_user_attributes` is also created so that each logged-in user
+	//   can see their own attributes.
+	// - MariaDB cannot `SET DEFAULT ROLE ALL`. To workaround this, a role
+	//   `tp-role-<username>` is created and assigned to the database user
+	//   while all roles are assigned to this all-in-one role. Then `SET
+	//   DEFAULT ROLE tp-role-<username>` before each session.
+	//
+	// Other MariaDB quirks:
+	// - In order to be able to grant a role, the grantor doing so must have
+	//   permission to do so (see WITH ADMIN in the CREATE ROLE article).
+	//   (Quoted from https://mariadb.com/kb/en/grant/#roles)
+	// - To set a default role for another user one needs to have write access
+	//   to the mysql database.
+	//   (Quoted from https://mariadb.com/kb/en/set-default-role/)
```

Comment:
Nit: I wonder if this comment wouldn't have a better place in the activate procedure itself.

Reply:
I was looking for a different place for this but the above go functions are shared for both MySQL/MariaDB. so I decided to keep the comments here.

Reply:
Fair enough.